File: 진우/29weeks/1991. 트리 순회/트리 순회.py
import sys
input = sys.stdin.readline

# -----------------------------------
# 각 노드의 왼쪽, 오른쪽 자식을 저장
# node['A'] = ('B', 'C') 형태
# -----------------------------------
node = {}

# -----------------------------
# 전위 순회 (Preorder)
# 순서: 루트 → 왼쪽 → 오른쪽
# -----------------------------
def preorder(n):
    if n == '.':  # 자식이 없는 경우 종료
        return
    
    print(n, end='')           # 1. 루트 출력
    preorder(node[n][0])       # 2. 왼쪽 자식
    preorder(node[n][1])       # 3. 오른쪽 자식


# -----------------------------
# 중위 순회 (Inorder)
# 순서: 왼쪽 → 루트 → 오른쪽
# -----------------------------
def inorder(n):
    if n == '.':
        return
    
    inorder(node[n][0])        # 1. 왼쪽 자식
    print(n, end='')           # 2. 루트 출력
    inorder(node[n][1])        # 3. 오른쪽 자식


# -----------------------------
# 후위 순회 (Postorder)
# 순서: 왼쪽 → 오른쪽 → 루트
# -----------------------------
def postorder(n):
    if n == '.':
        return
    
    postorder(node[n][0])      # 1. 왼쪽 자식
    postorder(node[n][1])      # 2. 오른쪽 자식
    print(n, end='')           # 3. 루트 출력


# -----------------------------------
# 입력
# -----------------------------------
N = int(input())

for _ in range(N):
    parent, left, right = input().split()
    node[parent] = (left, right)  # 딕셔너리에 저장

# -----------------------------------
# 순회 실행 (루트는 'A')
# -----------------------------------
preorder('A')
print()

inorder('A')
print()

postorder('A')

File: 진우/29weeks/1991. 트리 순회/트리 순회.cc
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

// 각 노드의 왼쪽, 오른쪽 자식을 저장
// node[0] → 'A', node[1] → 'B' ...
pair<char, char> node[26];


// -----------------------------
// 전위 순회 (Preorder)
// 순서: 루트 → 왼쪽 → 오른쪽
// -----------------------------
void preorder(char n) {
    // '.'은 자식이 없다는 의미 → 종료
    if (n == '.') {
        return;
    }

    cout << n; // 1. 루트 출력
    preorder(node[n - 'A'].first);   // 2. 왼쪽 자식
    preorder(node[n - 'A'].second);  // 3. 오른쪽 자식
}


// -----------------------------
// 중위 순회 (Inorder)
// 순서: 왼쪽 → 루트 → 오른쪽
// -----------------------------
void inorder(char n) {
    if (n == '.') {
        return;
    }

    inorder(node[n - 'A'].first);    // 1. 왼쪽 자식
    cout << n;                       // 2. 루트 출력
    inorder(node[n - 'A'].second);   // 3. 오른쪽 자식
}


// -----------------------------
// 후위 순회 (Postorder)
// 순서: 왼쪽 → 오른쪽 → 루트
// -----------------------------
void postorder(char n) {
    if (n == '.') {
        return;
    }

    postorder(node[n - 'A'].first);  // 1. 왼쪽 자식
    postorder(node[n - 'A'].second); // 2. 오른쪽 자식
    cout << n;                       // 3. 루트 출력
}


int main() {
    // 입출력 속도 향상
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int N; // 노드 개수
    cin >> N;

    // 트리 정보 입력
    for (int i = 0; i < N; i++) {
        char parent, left, right;
        cin >> parent >> left >> right;

        // parent를 인덱스로 변환해서 저장
        // 'A' → 0, 'B' → 1 ...
        node[parent - 'A'].first = left;   // 왼쪽 자식
        node[parent - 'A'].second = right; // 오른쪽 자식
    }

    // 루트는 항상 'A'로 시작
    preorder('A');   // 전위 순회
    cout << "\n";

    inorder('A');    // 중위 순회
    cout << "\n";

    postorder('A');  // 후위 순회

    return 0;
}